File: web/src/types/settlement.ts
export type SettlementStatus = 'pending' | 'partial' | 'settled' | 'disputed'

export interface SettlementPreview {
    settlement_id: number
    status: SettlementStatus
    office_name: string
    expected_total_cash: number
    expenses_to_reimburse: number
    net_cash_expected: number
    agency_presale_total: number
    outstanding_cargo_delivery: number
}

export interface Settlement {
    id: number
    trip: number
    trip_status: string
    office: number
    office_name: string
    conductor: number
    conductor_name: string
    settled_by: number | null
    settled_by_name: string | null
    origin_name: string
    destination_name: string
    expected_passenger_cash: number
    expected_cargo_cash: number
    expected_total_cash: number
    agency_presale_total: number
    outstanding_cargo_delivery: number
    expenses_to_reimburse: number
    net_cash_expected: number
    actual_cash_received: number | null
    actual_expenses_reimbursed: number
    discrepancy_amount: number | null
    reimbursement_gap: number
    status: SettlementStatus
    notes: string
    dispute_reason: string
    calculation_snapshot: Record<string, unknown>
    created_at: string
    updated_at: string
    settled_at: string | null
    disputed_at: string | null
    resolved_at: string | null
}

export interface SettlementListItem {
    id: number
    trip: number
    office: number
    office_name: string
    conductor: number
    conductor_name: string
    origin_name: string
    destination_name: string
    status: SettlementStatus
    expected_total_cash: number
    expenses_to_reimburse: number
    net_cash_expected: number
    actual_cash_received: number | null
    actual_expenses_reimbursed: number
    discrepancy_amount: number | null
    created_at: string
    settled_at: string | null
    disputed_at: string | null
    resolved_at: string | null
}

export interface PaginatedResponse<T> {
    count: number
    next: string | null
    previous: string | null
    results: T[]
}

File: backend/api/serializers/settlement.py
from rest_framework import serializers

from api.models import Settlement
from api.services.settlements import build_settlement_preview


class SettlementSerializer(serializers.ModelSerializer):
    office_name = serializers.CharField(source='office.name', read_only=True)
    conductor_name = serializers.CharField(source='conductor.get_full_name', read_only=True)
    settled_by_name = serializers.CharField(source='settled_by.get_full_name', read_only=True)
    origin_name = serializers.CharField(source='trip.origin_office.name', read_only=True)
    destination_name = serializers.CharField(source='trip.destination_office.name', read_only=True)
    trip_status = serializers.CharField(source='trip.status', read_only=True)
    reimbursement_gap = serializers.SerializerMethodField()

    class Meta:
        model = Settlement
        fields = [
            'id', 'trip', 'trip_status',
            'office', 'office_name',
            'conductor', 'conductor_name',
            'settled_by', 'settled_by_name',
            'origin_name', 'destination_name',
            'expected_passenger_cash', 'expected_cargo_cash',
            'expected_total_cash', 'agency_presale_total',
            'outstanding_cargo_delivery', 'expenses_to_reimburse',
            'net_cash_expected',
            'actual_cash_received', 'actual_expenses_reimbursed',
            'discrepancy_amount', 'reimbursement_gap',
            'status', 'notes', 'dispute_reason',
            'calculation_snapshot',
            'created_at', 'updated_at', 'settled_at', 'disputed_at', 'resolved_at',
        ]
        read_only_fields = fields

    def get_reimbursement_gap(self, obj):
        return obj.actual_expenses_reimbursed - obj.expenses_to_reimburse


class SettlementListSerializer(serializers.ModelSerializer):
    office_name = serializers.CharField(source='office.name', read_only=True)
    conductor_name = serializers.CharField(source='conductor.get_full_name', read_only=True)
    origin_name = serializers.CharField(source='trip.origin_office.name', read_only=True)
    destination_name = serializers.CharField(source='trip.destination_office.name', read_only=True)

    class Meta:
        model = Settlement
        fields = [
            'id', 'trip', 'office', 'office_name',
            'conductor', 'conductor_name',
            'origin_name', 'destination_name',
            'status', 'expected_total_cash', 'expenses_to_reimburse',
            'net_cash_expected', 'actual_cash_received',
            'actual_expenses_reimbursed', 'discrepancy_amount',
            'created_at', 'settled_at', 'disputed_at', 'resolved_at',
        ]
        read_only_fields = fields


class SettlementRecordSerializer(serializers.Serializer):
    actual_cash_received = serializers.IntegerField(min_value=0)
    actual_expenses_reimbursed = serializers.IntegerField(
        min_value=0, required=False, default=0,
    )
    notes = serializers.CharField(required=False, allow_blank=True, allow_null=True)


class SettlementDisputeSerializer(serializers.Serializer):
    dispute_reason = serializers.CharField(required=True, allow_blank=False)
    notes = serializers.CharField(required=False, allow_blank=True, allow_null=True)


class SettlementResolveSerializer(serializers.Serializer):
    actual_cash_received = serializers.IntegerField(min_value=0, required=False)
    actual_expenses_reimbursed = serializers.IntegerField(min_value=0, required=False)
    notes = serializers.CharField(required=True, allow_blank=False)


class SettlementPreviewSerializer(serializers.Serializer):
    settlement_id = serializers.IntegerField()
    status = serializers.CharField()
    office_name = serializers.CharField()
    expected_total_cash = serializers.IntegerField()
    expenses_to_reimburse = serializers.IntegerField()
    net_cash_expected = serializers.IntegerField()
    agency_presale_total = serializers.IntegerField()
    outstanding_cargo_delivery = serializers.IntegerField()

    @classmethod
    def from_settlement(cls, settlement):
        return cls(build_settlement_preview(settlement)).data
